internal/net/grpc/stats/stats.go

```diff
@@ -0,0 +1,394 @@
+				CpuLimitCores:    0,
+				CpuUsageCores:    0,
+				MemoryLimitBytes: 0,
+				MemoryUsageBytes: 0,
+			},
+		}
+		err = nil
+		return stats, err
+	}
+
+	stats = &payload.Info_ResourceStats{
+		Name: hostname,
+		Ip:   ip,
+		CgroupStats: &payload.Info_CgroupStats{
+			CpuLimitCores:    cgroupStats.CpuLimitCores,
+			CpuUsageCores:    cgroupStats.CpuUsageCores,
+			MemoryLimitBytes: cgroupStats.MemoryLimitBytes,
+			MemoryUsageBytes: cgroupStats.MemoryUsageBytes,
+		},
+	}
+
+	return stats, nil
+}
+
+// measureCgroupStats orchestrates the process of sampling and calculating cgroup statistics.
+func measureCgroupStats() (stats *CgroupStats, err error) {
+	// First sample: Read initial metrics from cgroup files (includes cumulative CPU usage)
+	m1, err := readCgroupMetrics()
+	if err != nil {
+		return nil, errors.ErrCgroupFirstSampleFailed(err)
+	}
+	t1 := time.Now()
+
+	// Wait 100ms to allow meaningful CPU usage accumulation
+	time.Sleep(100 * time.Millisecond)
+
+	// Second sample: Read metrics again to calculate CPU usage rate
+	m2, err := readCgroupMetrics()
+	if err != nil {
+		return nil, errors.ErrCgroupSecondSampleFailed(err)
+	}
+	t2 := time.Now()
+
+	// Calculate CPU usage rate from cumulative values: (usage2 - usage1) / time_delta
+	stats = calculateCpuUsageCores(m1, m2, t2.Sub(t1))
+
+	return stats, nil
+}
+
+// readCgroupMetrics reads raw memory & CPU metrics depending on cgroup mode
+func readCgroupMetrics() (metrics *CgroupMetrics, err error) {
+	mode := detectCgroupMode()
+	switch mode {
+	case CGV2:
+		metrics, err = readCgroupV2Metrics()
+	case CGV1:
+		metrics, err = readCgroupV1Metrics()
+	default:
+		err = errors.ErrCgroupModeDetectionFailed()
+	}
+
+	return metrics, err
+}
+
+// detectCgroupMode inspects /sys/fs/cgroup to detect cgroups mode
+func detectCgroupMode() CgroupMode {
+	// cgroups v2 unified mount has cgroup.controllers
+	if file.Exists(file.Join(cgroupBasePath, "cgroup.controllers")) {
+		return CGV2
+	}
+
+	data, err := file.ReadFile("/proc/self/cgroup")
+	if err != nil {
+		return Unknown
+	}
+	lines := strings.Split(string(data), "\n")
+	for _, line := range lines {
+		// v2 line looks like: "0::/kubepods.slice/..."
+		parts := strings.Split(line, ":")
+		if len(parts) >= 3 && parts[1] == "" {
+			return CGV2
+		}
+	}
+
+	// If not v2, assume v1 on most distros
+	return CGV1
+}
+
+// readCgroupV2Metrics reads cgroups v2 raw metrics
+func readCgroupV2Metrics() (metrics *CgroupMetrics, err error) {
+	data, err := file.ReadFile(file.Join(cgroupBasePath, "memory.current"))
+	if err != nil {
+		return nil, errors.ErrCgroupV2MemoryCurrentRead(err)
+	}
+	memCur, err := strconv.ParseUint(strings.TrimSpace(string(data)), 10, 64)
+	if err != nil {
+		return nil, errors.ErrCgroupV2MemoryCurrentParse(err)
+	}
+
+	data, err = file.ReadFile(file.Join(cgroupBasePath, "memory.max"))
+	if err != nil {
+		return nil, errors.ErrCgroupV2MemoryMaxRead(err)
+	}
+	memMaxStr := strings.TrimSpace(string(data))
```

Comment:
You can use internal/conv.Btoa here
```suggestion
	memMaxStr := strings.TrimSpace(conv.Btoa(data))
```

Reply:
fixed in commit 118351ebd72822d82d7c7929dc2ad59f2583872d